File: java/src/org/trieck/pixievm/Memory.java
package org.trieck.pixievm;

import java.io.IOException;
import java.io.InputStream;

public class Memory {

    private static final int MEM_SIZE = 0x10000;
    private byte[] memory;
    private static Memory instance = null;

    /**
     * Public ctor
     */
    private Memory() {
        memory = new byte[MEM_SIZE];
    }

    /**
     * Singleton instance
     */
    public static Memory getInstance() {
        if (instance == null)
            instance = new Memory();
        return instance;
    }

    /**
     * Load stream into memory
     *
     * @param is
     * @param address
     * @param len
     * @return number of bytes read
     * @throws IOException
     */
    int load(InputStream is, int address, int len) throws IOException {
        return is.read(memory, address, len);
    }

    /**
     * Fetch a byte
     *
     * @param address
     * @return byte @ address
     */
    byte fetch(int address) {
        return memory[address & 0x0000FFFF];
    }

    /**
     * Store a byte
     *
     * @param address
     * @param b       byte
     */
    void store(int address, byte b) {
        memory[address & 0x0000FFFF] = b;
    }
}

File: c++/pixievm/pixievm/Memory.cpp
/////////////////////////////////////////////////////////////////////////////
//
// MEMORY.CPP : Pixie Virtual Machine Memory
//
// Copyright (c) 2006-2009, Thomas A. Rieck, All Rights Reserved
//

#include "Common.h"
#include "Memory.h"

MemoryPtr Memory::instance(Memory::getInstance());

/////////////////////////////////////////////////////////////////////////////
Memory::Memory()
{
	memory = new byte[MEM_SIZE];
	memset(memory, 0, MEM_SIZE * sizeof(byte));
}

/////////////////////////////////////////////////////////////////////////////
Memory::~Memory()
{
	delete[] memory;
}

/////////////////////////////////////////////////////////////////////////////
Memory *Memory::getInstance()
{
	if (instance.get() == NULL) {
		instance = MemoryPtr(new Memory);
	}
	return instance.get();
}

/////////////////////////////////////////////////////////////////////////////
byte Memory::fetch(word address)
{
	return memory[address & 0x0000FFFF];
}

/////////////////////////////////////////////////////////////////////////////
void Memory::store(word address, byte b)
{
	memory[address & 0x0000FFFF] = b;
}

/////////////////////////////////////////////////////////////////////////////
bool Memory::load(istream & is, word base, int size)
{
	is.read((char *) &memory[base], size);
	return is.good();
}

/////////////////////////////////////////////////////////////////////////////
bool Memory::save(ostream & os, word base, int size)
{
	os.write((const char *)&memory[base], size);
	return os.good();
}